### ISSUE TITLE
активация однократным нажатием

### ISSUE THREAD

**Issue by ivanbalashov239** [uzyovoys/aggregate#36]
На данный момент любая активация должна сама сообщить когда пора перестать слушать.
Если активировать распознавание по отжатию клавиши(комбинации), то останавливать его можно когда человек перестал говорить. То есть слушать микрофон отдельным процессом и посылать сигнал когда он молчит какое то время. Это можно сделать с помощью [sox](http://sox.sourceforge.net/), 
`sox -d /tmp/record.flac silence 1 0.1 5% 1 2.8 5%`
 например здесь после того как человек на 2.8 секунд замолчит(громкость меньше 5%) процесс записи остановится, соответственно можно послать сигнал завершить запись распознавания. Наверное можно сделать более красиво, используя эту библиотеку, но я не нашел обертки для джавы этой либы.
Как поступим? Что используется в Дусе?

**Comment by morfeusys** [comment]
@seniorivn естественно. Это нужно и для задачи #2 
Но с точки зрения алгоритма - там немного не в громкости дело. Самый простой алгоритм - алгоритм гейта. Нужно на старте записи замерять короткое время уровень сигнал-шум, а потом отслеживать, когда он падает ниже этого порога и срабатывать через нужный таймаут.

Это должно быть системной функцией самого Агрегата, поэтому беру на себя. Кроме того, это повлечет введение нового типа события типа asr.listen, который будет обрабатываться самим Агрегатом.

**Comment by ivanbalashov239** [comment]
@Morfeusys  ок, тогда пусть повисит, пока агрегат не обновишь. Тогда key-detector и обновлю.

**Comment by morfeusys** [comment]
@seniorivn единственная проблема с активацией по однократному нажатию клавиши клавиатуры в том, что это потенциально опасное действие. Можно назначить одну клавишу и случайно ее нажать...

**Comment by ivanbalashov239** [comment]
@Morfeusys ии? пользователь имеет право стрелять себе в ногу :)
Можно из вежливости предупредить, что не надо так делать с используемыми в быту клавишами.
Ну а когда будет графические настройки это вообще будет не проблемой.

**Comment by morfeusys** [comment]
@seniorivn я не к тому что этого нельзя делать. Просто тогда это накладывает сопутствующие требования. Такие как сигнал начала и окончания распознавания. Иначе обратной связи нет.

**Comment by ivanbalashov239** [comment]
@Morfeusys  ещё вариант, запускать однократным нажатием запись, а останавливать при нажатии любой клавиши. Это я и сейчас могу сделать.

**Comment by morfeusys** [comment]
@seniorivn не подойдет как общее решение, так как во время говорения я могу печатать что-то в другой программе, как я собственно часто и делаю.

**Comment by ivanbalashov239** [comment]
@Morfeusys да, ну это же костыли, пока нет автоматически завершающегося процесса распознавания. Сделаете его, будем в этой части двигаться дальше, а пока это узкое место.
